File: backend/src/main/java/com/anushibinj/mypromptlib/service/GoogleTokenVerifierService.java
package com.anushibinj.mypromptlib.service;

import lombok.AllArgsConstructor;
import lombok.Data;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Service
public class GoogleTokenVerifierService {

    private final RestTemplate restTemplate;
    private final String clientId;

    public GoogleTokenVerifierService(RestTemplate restTemplate, @Value("${google.client-id}") String clientId) {
        this.restTemplate = restTemplate;
        this.clientId = clientId;
    }

    @SuppressWarnings("unchecked")
    public GoogleUserInfo verify(String idToken) {
        String url = "https://oauth2.googleapis.com/tokeninfo?id_token=" + idToken;
        Map<String, Object> response;
        try {
            response = restTemplate.getForObject(url, Map.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to verify Google token");
        }
        if (response == null) {
            throw new RuntimeException("Invalid Google token");
        }
        String aud = (String) response.get("aud");
        if (!clientId.equals(aud)) {
            throw new RuntimeException("Token not intended for this app");
        }
        String email = (String) response.get("email");
        String sub = (String) response.get("sub");
        if (email == null || sub == null) {
            throw new RuntimeException("Invalid Google token: missing user info");
        }
        return new GoogleUserInfo(sub, email);
    }

    @Data
    @AllArgsConstructor
    public static class GoogleUserInfo {
        private String googleId;
        private String email;
    }
}

File: backend-go/internal/service/google_token_verifier_service.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
)

type GoogleTokenVerifierService struct {
	clientID string
}

func NewGoogleTokenVerifierService(clientID string) *GoogleTokenVerifierService {
	return &GoogleTokenVerifierService{clientID: clientID}
}

type GoogleUserInfo struct {
	GoogleID string
	Email    string
}

func (s *GoogleTokenVerifierService) Verify(idToken string) (*GoogleUserInfo, error) {
	url := fmt.Sprintf("https://oauth2.googleapis.com/tokeninfo?id_token=%s", idToken)

	resp, err := http.Get(url)
	if err != nil {
		return nil, errors.New("Failed to verify Google token")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.New("Invalid Google token")
	}

	var responseMap map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&responseMap); err != nil {
		return nil, errors.New("Invalid Google token")
	}

	aud, ok := responseMap["aud"].(string)
	if !ok || aud != s.clientID {
		return nil, errors.New("Token not intended for this app")
	}

	email, ok1 := responseMap["email"].(string)
	sub, ok2 := responseMap["sub"].(string)

	if !ok1 || !ok2 || email == "" || sub == "" {
		return nil, errors.New("Invalid Google token: missing user info")
	}

	return &GoogleUserInfo{
		GoogleID: sub,
		Email:    email,
	}, nil
}
